### PR TITLE
[framework] remove unnecessary abilities

### PR DESCRIPTION
**Pull request by lightmark**
### Description

As discussed. Let's remove unnecessary abilities.

tldr:
1. We remove `store` ability constraint on most data structures including simple_map to enable ephemeral usage of those structs. For example, then you can create a temporary simple_map to dedup (K, V)s where K or V is not `store`.
2. We remove the unnecessary generic constraint on all methods except for the constructor such as what we did for table is. This is to make it more consistent with the pattern on other data structures where we don't annotate type arguments with constraints if not needed since constructor functions will enforce that anyway. We want to stress this will be a pattern that we will follow along the way.

### Test Plan
cargo test

### CODE REVIEW

**Comment by lightmark** [comment]
`tests::framework_compatibility::can_upgrade_framework_on_testnet ... ok`?

**Review by movekevin — COMMENTED** [review]
Can you add more context to the PR's description to explain these removals? For example, we discussed removing store constraint on type arguments in simple_map so it can be used for ephemeral storage (e.g. deduplicating inputs)

**Comment by lightmark** [comment]
> > I reckon that the compatibility checker might fail on simple map. have you validated? the logic for abilities is that it must be a superset in upgrades it cannot be a reduced set.
> 
> I think you're thinking of struct abilities. Removing type arguments constraints is allowed by compatibility check - the new ability constraints needs to be a subset (not superset): https://github.com/move-language/move/blob/main/language/move-binary-format/src/compatibility.rs#L252
> 
> For struct abilities, the new abilities need to be a superset: https://github.com/move-language/move/blob/main/language/move-binary-format/src/compatibility.rs#L213

This doesn't say anything about type constraint on functions?

**Comment by movekevin** [comment]
> > > I reckon that the compatibility checker might fail on simple map. have you validated? the logic for abilities is that it must be a superset in upgrades it cannot be a reduced set.
> > 
> > 
> > I think you're thinking of struct abilities. Removing type arguments constraints is allowed by compatibility check - the new ability constraints needs to be a subset (not superset): https://github.com/move-language/move/blob/main/language/move-binary-format/src/compatibility.rs#L252
> > For struct abilities, the new abilities need to be a superset: https://github.com/move-language/move/blob/main/language/move-binary-format/src/compatibility.rs#L213
> 
> This doesn't say anything about type constraint on functions?

By type argument constraints, I was referring to function type constraints (type arguments)

**Comment by lightmark** [comment]
> > > > I reckon that the compatibility checker might fail on simple map. have you validated? the logic for abilities is that it must be a superset in upgrades it cannot be a reduced set.
> > > 
> > > 
> > > I think you're thinking of struct abilities. Removing type arguments constraints is allowed by compatibility check - the new ability constraints needs to be a subset (not superset): https://github.com/move-language/move/blob/main/language/move-binary-format/src/compatibility.rs#L252
> > > For struct abilities, the new abilities need to be a superset: https://github.com/move-language/move/blob/main/language/move-binary-format/src/compatibility.rs#L213
> > 
> > 
> > This doesn't say anything about type constraint on functions?
> 
> By type argument constraints, I was referring to function type constraints (type arguments)

What about struct type parameter constraints? The same?

**Comment by movekevin** [comment]
> > > > > I reckon that the compatibility checker might fail on simple map. have you validated? the logic for abilities is that it must be a superset in upgrades it cannot be a reduced set.
> > > > 
> > > > 
> > > > I think you're thinking of struct abilities. Removing type arguments constraints is allowed by compatibility check - the new ability constraints needs to be a subset (not superset): https://github.com/move-language/move/blob/main/language/move-binary-format/src/compatibility.rs#L252
> > > > For struct abilities, the new abilities need to be a superset: https://github.com/move-language/move/blob/main/language/move-binary-format/src/compatibility.rs#L213
> > > 
> > > 
> > > This doesn't say anything about type constraint on functions?
> > 
> > 
> > By type argument constraints, I was referring to function type constraints (type arguments)
> 
> What about struct type parameter constraints? The same?

"struct type parameter constraints" = struct abilities in my original comment. When you say struct A has key. key is not a constraint. It's an ability.

File: aptos-move/framework/aptos-stdlib/sources/table.move [hunk 1]
```diff
@@ -9,7 +9,7 @@ module aptos_std::table {
     friend aptos_std::table_with_length;
 
     /// Type of tables
-    struct Table<phantom K: copy + drop, phantom V> has store {
+    struct Table<phantom K, phantom V> has store {
```
Comment (areshand) [inline aptos-move/framework/aptos-stdlib/sources/table.move:12]: Sometimes, the constraint is a way to help users. is there any scenario that copy or drop not required? 
For example, if I accidentally use a struct (for example, token) that is not droppable as a key, I later maynot be able to remove the item from the table since I cannot destroy it.

Comment (movekevin) [reply aptos-move/framework/aptos-stdlib/sources/table.move:12]: The copy/drop constraints are already enforced on creation so technically even if the struct def doesn't enforce it, it's not possible to create a table with non-copyable non-droppable keys.

Comment (areshand) [reply aptos-move/framework/aptos-stdlib/sources/table.move:12]: Ok, we still need these constraints. Only ability at the struct level is redundant since the function already enforce these constraints.

Comment (lightmark) [reply aptos-move/framework/aptos-stdlib/sources/table.move:12]: From Rust perspective, it doesn't have to be copy + drop. Even if we don't enforce on creation, users will find it is infeasible to use in their case by the api we provide so it won't be a technical issue.

**Comment by davidiw** [comment]
Let's get a PR that passes ci/cd. I'm happy to see some clarity here, if the compiler doesn't enforce something, then we should probably lean towards flexible unless it is well documented.

**Review by movekevin — APPROVED** [review]

File: aptos-move/move-examples/data_structures/sources/bucket_table.move [hunk 2]
```diff
@@ -37,7 +37,7 @@ module aptos_std::bucket_table {
     }
 
     /// Create an empty BucketTable with `initial_buckets` buckets.
-    public fun new<K: drop + store, V: store>(initial_buckets: u64): BucketTable<K, V> {
+    public fun new<K: copy + store + drop, V: store>(initial_buckets: u64): BucketTable<K, V> {
```
Comment (areshand) [inline aptos-move/move-examples/data_structures/sources/bucket_table.move:40]: why not just new<K, V> here?

File: aptos-move/move-examples/data_structures/sources/bucket_table.move [hunk 3]
```diff
@@ -144,7 +144,7 @@ module aptos_std::bucket_table {
     /// Aborts if there is no entry for `key`.
     /// The requirement of &mut BucketTable is to bypass the borrow checker issue described in https://github.com/move-language/move/issues/95
     /// Once Table supports borrow by K, we can remove the &mut
-    public fun borrow<K: copy + drop, V>(map: &mut BucketTable<K, V>, key: K): &V {
+    public fun borrow<K: drop, V>(map: &mut BucketTable<K, V>, key: K): &V {
```
Comment (areshand) [inline aptos-move/move-examples/data_structures/sources/bucket_table.move:147]: why #40 requires copy and here no copy?

**Review by davidiw — APPROVED** [review]

File: aptos-move/move-examples/data_structures/sources/bucket_table.move (hunk 2 above)
Comment (lightmark) [reply aptos-move/move-examples/data_structures/sources/bucket_table.move:40]: As discussed, we will enforce the abilities only at creation.

File: aptos-move/move-examples/data_structures/sources/bucket_table.move (hunk 3 above)
Comment (lightmark) [reply aptos-move/move-examples/data_structures/sources/bucket_table.move:147]: https://github.com/aptos-labs/aptos-core/pull/6339
